components/layout/dom_traversal.rs

```diff
@@ -219,6 +221,16 @@ fn traverse_children_of<'dom>(
         } else {
             handler.handle_text(&info, node_text_content);
         }
+    } else if parent_element.is_text_editing_root() {
+        let info = NodeAndStyleInfo::new(
+            parent_element,
+            parent_element.style(context.shared_context()),
+        );
+        for child in iter_child_nodes(parent_element) {
+            if child.is_text_node() {
+                handler.handle_text(&info, child.to_threadsafe().node_text_content());
+            }
+        }
```

Comment:
Can you explain why it is necessary to skip non-text children of text editing roots? It looks like in the current version of the shadow tree there aren't even any other children there?

Reply:
In terms of text input, it would be incorrect for a text editing root to contain other type of nodes. I suppose I could put a debug log here.

Reply:
Since we have full control over what is inside the shadow tree I don't think we should modify `dom_traversal.rs` at all.

Reply:
I guess we still need to handle this kind of style context passing. Since the current implementation of caret and selection depends on parent's style. This part of the code will actually replace the "input" element specific if condition above. Which, also handle empty input text.